### DISCUSSION TITLE
docfx v3 - Newtonsoft.Json

### DISCUSSION THREAD

**Discussion by VaclavElias** [dotnet/docfx#8137] [Q&A] [answered]
Any reason you are using ```Newtonsoft.Json```? It seems to be used only in one method. Any reason not to use ```System.Text.Json```?

https://github.com/dotnet/docfx/blob/0bfc8b7f60fff7c56423e97575104783151472a2/src/Microsoft.Docs.MarkdigExtensions/CodeSnippet/HtmlCodeSnippetRenderer.cs#L307-L346

**Comment by KalleOlaviNiemitalo** [comment]
AFAIK, System.Text.Json does not implement JSON Path used in the <code>contentObject.[SelectToken](https://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_Linq_JToken_SelectToken.htm)(sourceJsonPath)</code> call. I don't see such a method in [JsonElement Struct](https://docs.microsoft.com/dotnet/api/system.text.json.jsonelement?view=net-6.0).

**Reply by KalleOlaviNiemitalo** [reply]
<https://github.com/dotnet/runtime/issues/31068> is a request to implement JsonPath queries in System.Text.Json, quite popular, but not going to be in .NET 7.

**Reply by paulushub** [reply]
If I understand his concern well as expressed in **It seems to be used only in one method**, this is a specific query and does not really require that much of JSON support to implement.
Moreover, most .NET documentation projects do not need or use Notebook support.

**Reply by KalleOlaviNiemitalo** [reply]
`git grep -l -F 'Newtonsoft.Json' 3.0.0-beta1.1235+f5a62adcac -- "src/*.cs" | wc -l` finds 116 files; the same for System.Text.Json finds 5 files. Before <https://github.com/dotnet/docfx/pull/7773>, the numbers were 119 and 4. So, there has been some progress towards System.Text.Json but Newtonsoft.Json still has the majority. If the notebook feature is rarely used, I suppose porting it is not important for performance.

**Reply by VaclavElias** [reply]
> `git grep -l -F 'Newtonsoft.Json' 3.0.0-beta1.1235+f5a62adcac -- "src/*.cs" | wc -l` finds 116 files; the same for System.Text.Json finds 5 files. Before #7773, the numbers were 119 and 4. So, there has been some progress towards System.Text.Json but Newtonsoft.Json still has the majority. If the notebook feature is rarely used, I suppose porting it is not important for performance.

Thanks. It seems I should improve my search capabilities 🤦‍♂️🤣. I was just surprised that v3 being so long in development is still using ```Newtonsoft.Json``` but as you pointed out, there are some blockers.

Thanks for your comments everyone.

**Comment by yufeih** [comment, marked as answer]
When we were building the JSON pipeline for v3, `System.Text.Json` was very immature. 

Today the biggest blocker to move to `System.Text.Json` is the propagation of line info. We rely on this propagation mechanism to report detailed diagnostic locations for YAML files. With `Newtonsoft.Json`, we can [attach custom source info](https://github.com/dotnet/docfx/blob/v3/src/docfx/lib/json/JsonUtility.cs#L476) to `JToken` and deserialize it into [`Source<T>`](https://github.com/dotnet/docfx/blob/v3/src/docfx/lib/log/SourceInfo%7BT%7D.cs) with a [custom JsonConverter](https://github.com/dotnet/docfx/blob/v3/src/docfx/lib/json/SourceInfoJsonConverter.cs). There doesn't seem to be a straightforward way of achieving the same in `System.Text.Json`.

**Reply by paulushub** [reply]
> Today the biggest blocker to move to System.Text.Json is the propagation of line info.

Too bad, many efforts were made but @layomia did not see the need for this still - it seems MS needs developers with real-world experience to manage its products.

https://github.com/dotnet/runtime/issues/50629
https://github.com/dotnet/runtime/issues/47689
